### PR TITLE
[IMP] website: better vertical align and horizontal spacing in systray

### PR DESCRIPTION
**Pull request by rdeodoo**
Before this commit, both the "Edit" (in main lang) and "Translate" (in
alternate lang) button in the website navbar had a vertical alignment
issue.

This was because, despite having the same font style and font size, they
had a different line-height making it glitch 1-2px vertically.

It's a bit less visible for the Edit button as there is a pencil icon
between this text and the text next to it.
For the translate button, it's quite visible directly.

For testing purpose, you can simply remove the icons and the left-right
margins of those systray items so the text are glued to each other and
you clearly see the misindentation.

For tracking purpose, there is multiple cases to be tested regarding
those buttons:
- Community vs Enterprise (where those are colored)
- Regular mode vs translate mode

| Use Case | Issue + Screenshot |
| -- | -- |
| **Regular Mode CE** | Not enough horizontal spacing between items + **"Edit" text is not vertically align**|
| Before |  image  |
| After | image |
| **Regular Mode EE** | Not enough horizontal spacing between items |
| Before |  image |
| After | image |
| **Translate Mode CE** | Not enough horizontal spacing between items + **"Translate" text is not vertically align** |
| Before |  image |
| After | image|
| **Translate Mode EE** | Not enough horizontal spacing between items |
| Before |  image|
| After |image |

### CODE REVIEW

**Comment by rdeodoo** [comment]
@stefanorigano See https://github.com/odoo/odoo/pull/130243#issuecomment-1669499933
Note that I did not have time to fully test it (the valign fix), in local it works well but I wanted to battle test it on runbot.
If you add your spacing fix, this PR should be fixing every spacing and align regarding the website systray :+1:

**Comment by rdeodoo** [comment]
@qsm-odoo If you have a few minutes, this is a small one about visual nitpicking, but the community "Translate" button vertical misalignment was too much for me to keep ignoring it :D
And stefano actually pointed out that the horizontal spacing was also too narrow.
Feel free to delegate to someone

File: addons/website/static/src/components/navbar/navbar.scss [hunk 1]
```diff
@@ -0,0 +1,14 @@
+header {
+    .o_main_navbar {
```
Comment (stefanorigano) [inline addons/website/static/src/components/navbar/navbar.scss:2]: In general, I think that modules shouldn't target webclient classes. 
I'd use `.o_website_systray` as main parent selector instead.

**Review by stefanorigano — COMMENTED** [review]
Hello @rdeodoo 👋
I've left a couple of comments, but these may be are more about team's convention I guess, I let you guys decide what to do 👍

File: addons/website/static/src/components/navbar/navbar.scss [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        .o_website_systray {
+            --NavBar-entry-margin-right: #{$o-navbar-entry-padding-h};
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/components/navbar/navbar.scss:5]: I don't know why the web app is that generic about this but I doubt we should add any margin, only paddings. Especially not as a website exception. Worse: there are actually spaces *between* button which actually click on the button somehow:
- Hover the publish button -> you have an hover effect
- Go towards the mobile button -> you have a space between the two buttons where you don't have an hover effect
- Click -> the publish button is enabled :fearful: 

Note: I don't know if this is a bug that this PR creates, my intuition is that the PR *reveals* it adding margins with the generic `--NavBar-entry-margin-right` stuff. Maybe this requires a deeper framework-level fix.

File: addons/website/static/src/systray_items/translate_website.scss [hunk 3]
```diff
@@ -1,3 +1,3 @@
 .o_translate_website_container {
-    margin: -$o-navbar-padding-v 0;
+    margin: -$o-navbar-padding-v var(#{--NavBar-entry-margin-right}, 0) -$o-navbar-padding-v 0;
```
Comment (qsm-odoo) [inline addons/website/static/src/systray_items/translate_website.scss:2]: The `#{}` seems useless

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/components/navbar/navbar.scss [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+        @if (not $o-navbar-use-pills) {
+            .o_website_systray .o_menu_systray_item > .btn {
+                /* This line is extracted from `%-main-navbar-entry-spacing` */
+                line-height: calc(var(--o-navbar-height) - #{$o-navbar-padding-v * 2});
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss:10]: - I can't extend it, it comes with a lot of rules which we don't need sadly (and which breaks the layout)
- I don't think we should do anything in the web app, `o_menu_systray_item` is only website related? It was initially in the web app but @stefanorigano pointed out that it was a website only class, which I didn't realize, so I moved it to website with this [push force](https://github.com/odoo/odoo/compare/46b40997eaac9201acfedcfbbea4c1937d75ae63..8653f8f75dd6bde0d01a27443cdb6dbc3014c73f)

File: addons/website/static/src/components/navbar/navbar.scss (hunk 2 above)
Comment (stefanorigano) [reply addons/website/static/src/components/navbar/navbar.scss:5]: This is on me 🙋‍♂️ (sorry @rdeodoo ).
I've switched to enterprise and applied some changes, then I've forgot to test community again .

I'll fix this and https://github.com/odoo/odoo/pull/131228#discussion_r1296884756 as well

Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss:5]: Ok perfect, I was actually changing the selector and replying to the other QSM review but I'll let you do it :+1:

File: addons/website/static/src/components/navbar/navbar.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/components/navbar/navbar.scss:10]: Did not realize `o_menu_systray_item` was a website class either but my point was that the line-height should be set at framework level (it can be added on the navbar itself actually) and inherited by specific stuff (if website really need specific stuff).

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/website/static/src/components/navbar/navbar.scss:10]: > I don't think we should do anything in the web app, o_menu_systray_item is only website related? It was initially in the web app but @stefanorigano pointed out that it was a website only class, which I didn't realize, so I moved it to website with this [push force](https://github.com/odoo/odoo/compare/46b40997eaac9201acfedcfbbea4c1937d75ae63..8653f8f75dd6bde0d01a27443cdb6dbc3014c73f)

@rdeodoo ,  there is a misunderstanding :)
`o_menu_systray_item` it's **not** website only... now I do understand your answer on discord ^^

To clarify what I meant:  if you experience misalignments in website only, then the fix should be placed there. Otherwise a general fix is needed ofc.

I've got in touch because we discussed about a website's issue (translate button etc..) and I was initially surprised to see your fix in the main app, I wanted to double-check (again, I can't reproduce the issue).


---


> Jesus, that line is ugly

@qsm-odoo , what's annoying you here 😅 ? 
```scss
 line-height: calc(var(--o-navbar-height) - #{$o-navbar-padding-v * 2});
 ```
`--o-navbar-height` adapts according to the screen size* and `$o-navbar-padding-v` according to the version  (community uses `0px`, enterprise uses `10px` instead).

(*) since the very last _milk_ iteration the navbar height it's fixed, thus we could replace the CSS variable for a SCSS one, meaning no need of `calc()` anymore. I suspect that's 'cause you're not a fan of `calc()` in general (and I understand why)... but probably we'll have to live with it until either CSS provides a "better" way to handle variables interpolation or sass allows us to define variables into media queries (probably just not possible).
There are workarounds I know, but this seemed the cleanest solution to me 🤷🏻‍♂️



> in standard web app, o_menu_systray_item should have the correct line-height, then at worst this should be about adding a line-height: inherit on the inner button here.

Well, `line-height` is not the only property we have to tickle in order to achieve consistency across all the different types of entries. I found cleaner to extend pseudo-classes rather that trying to force specific values for each element. Now that all the code moved to OWL (and we finally reached a good level of consistency) we could simplify I guess, I'll have a look when possible.

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss:10]: @stefanorigano 

> To clarify what I meant: if you experience misalignments in website only, then the fix should be placed there. Otherwise a general fix is needed ofc.
> 
> I've got in touch because we discussed about a website's issue (translate button etc..) and I was initially surprised to see your fix in the main app, I wanted to double-check (again, I can't reproduce the issue).

Well, there is no case other than website for now, but if you create any `.o_menu_systray_item` element with a `.btn` inside in any systray in Odoo, it will have the same issue.
For instance, you can just copy this HTML inside the regular Odoo systray and see the vertical misalignment:
```
<a href="#" class="h-100 btn btn-primary d-flex align-items-center rounded-0 border-0 fw-normal" accesskey="a">
    <span class="fa me-2 fa-pencil"/>Edit
</a>
```
image

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/components/navbar/navbar.scss:10]: > > Jesus, that line is ugly
>
> @qsm-odoo , what's annoying you here 😅 ?

@stefanorigano Nothing annoys me, I said it was ugly not useless :wink: You can be ugly and needed, look at @rdeodoo :smile:

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss:10]: image

**Comment by stefanorigano** [comment]
@rdeodoo , I have fixed [these two](https://github.com/odoo/odoo/pull/131228#discussion_r1296943020)  and pushed a --fixup on top... it make sense to me but I need you to have a look,  I tried but I'm still not able to replicate 😅

**Comment by rdeodoo** [comment]
> I tried but I'm still not able to replicate 😅

It's not working, the vertical align is broken again :( see:

image

Note that you can simply remove the left/right margin/padding to better see it
image

I didn't check enterprise

**Comment by stefanorigano** [comment]
> It's not working, the vertical align is broken 

Oh no 🫣 Well, clearly I can't fix it without replicating the issue.
Btw, I still think that's system-fonts related...

> Do you think we can just remove the fixup commit and proceed to the merge phase or do you want to still work on the fixup commit and make it work?

Sure, go ahead. The only purpose of the fixup was to remove the ugly line, but we'll have to live with it I guess 👍

**Comment by rdeodoo** [comment]
@stefanorigano It's indeed linked to system font

File: addons/website/static/src/components/navbar/navbar.xml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+<t t-inherit="web.NavBar" t-inherit-mode="extension" owl="1">
+    <xpath expr="//div[hasclass('o_menu_systray')]" position="attributes">
+        <attribute name="t-att-class">isWebsiteSystray ? 'o_website_systray' : ''</attribute>
+    </xpath>
+</t>
+
+</templates>
```
Comment (qsm-odoo) [inline addons/website/static/src/components/navbar/navbar.xml:10]: Not a fan of this (if another app does the same, everything is broken) and still think it should not be necessary (there should not be website specific design here).

And if it was necessary, this should be squashed IMO.

File: addons/website/static/src/components/navbar/navbar.js [hunk 6]
```diff
@@ -44,11 +44,18 @@ patch(NavBar.prototype, 'website_navbar', {
         useBus(websiteSystrayRegistry, 'CONTENT-UPDATED', renderAndAdapt);
     },
 
+    get isWebsiteSystray() {
```
Comment (qsm-odoo) [inline addons/website/static/src/components/navbar/navbar.js:47]: Seems strange that a "navbar" object has a function "isWebsiteSystray"... should not it be something like "shouldDisplayWebsiteSystray", something like that?

File: addons/website/static/src/components/navbar/navbar.scss [hunk 7]
```diff
@@ -0,0 +1,8 @@
+.o_menu_systray.o_website_systray {
+    @if not $o-navbar-use-pills {
+        .o_menu_systray_item > .btn {
+            /* This line is extracted from `%-main-navbar-entry-spacing` */
+            line-height: calc(var(--o-navbar-height) - #{$o-navbar-padding-v * 2});
+        }
+    }
+}
```
Comment (qsm-odoo) [inline addons/website/static/src/components/navbar/navbar.scss]: I am still thinking the same thing as earlier: why is this necessary, why is this line-height not inherited instead of duplicated and why is this a necessary website override?

File: addons/website/static/src/components/navbar/navbar.scss [hunk 8]
```diff
@@ -1,5 +1,10 @@
 .o_menu_systray.o_website_systray {
-    @if not $o-navbar-use-pills {
+    @if $o-navbar-use-pills {
+        --NavBar-entry-margin-right: #{$o-navbar-entry-padding-h};
+    } @else {
+        --NavBar-entry-padding-left: #{$o-navbar-entry-padding-h};
+        --NavBar-entry-padding-right: #{$o-navbar-entry-padding-h};
+
```
Comment (qsm-odoo) [inline addons/website/static/src/components/navbar/navbar.scss:7]: Is this related to this or not: I still have the same bug in enterprise as I mentioned before. There is a gap between buttons where the cursor is like "you are not clicking on anything" but if you click, you are actually triggering the button (e.g. between publish button and mobile button).

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
This is surely better with this PR, but let's solve everything in one-go... no? :pray:

File: addons/website/static/src/systray_items/edit_website.xml [hunk 9]
```diff
@@ -1,11 +1,13 @@
...
 <templates xml:space="preserve">
 <t t-name="website.EditWebsiteSystray" owl="1">
-    <div class="o_edit_website_container d-none d-md-block" t-att-class="{ o_translatable: translatable }">
-        <a href="#" class="h-100" accesskey="a" t-on-click="startEdit" t-att-class="{'btn btn-primary d-flex align-items-center rounded-0 border-0 fw-normal': !translatable, 'o_nav_entry': translatable}">
-            <span t-if="translatable" class="fa fa-pencil me-2"/>
-            <span t-else="" class="fa me-2" t-att-class="{ 'fa-refresh fa-spin': this.state.isLoading, 'fa-pencil': !this.state.isLoading }"/>
-            <t t-esc="label"/>
+    <div class="o_menu_systray_item o_edit_website_container d-none d-md-block" t-att-class="{ o_translatable: translatable, o_not_translatable: !translatable }">
+        <a href="#" class="h-100" accesskey="a" t-on-click="startEdit" t-att-class="{'btn btn-primary d-flex align-items-center ps-0 pe-2 rounded-0 border-0 fw-normal': !translatable, 'o_nav_entry': translatable}">
+            <span  t-att-class="{'o_nav_entry' : !translatable}">
```
Comment (rdeodoo) [inline addons/website/static/src/systray_items/edit_website.xml:6]: To clean whenever someone edit the pr
```suggestion
            <span t-att-class="{'o_nav_entry' : !translatable}">
```

File: addons/website/static/src/systray_items/translate_website.xml [hunk 10]
```diff
@@ -2,10 +2,12 @@
 <templates xml:space="preserve">
 
     <t t-name="website.TranslateWebsiteSystray" owl="1">
-        <div class="o_translate_website_container d-none d-md-block">
-            <a href="#" accesskey="t" t-on-click="startTranslate" class="btn btn-primary d-flex align-items-center h-100 rounded-0 border-0 fw-normal">
-                <span t-if="websiteContext.translation" class="fa fa-refresh fa-spin me-2"/>
-                Translate
+        <div class="o_menu_systray_item o_translate_website_container d-none d-md-block">
+            <a href="#" accesskey="t" t-on-click="startTranslate" class="btn btn-primary d-flex align-items-center h-100 px-0 rounded-0 border-0 fw-normal">
+                <span class="o_nav_entry">
```
Comment (rdeodoo) [inline addons/website/static/src/systray_items/translate_website.xml:7]: So the new commit is adding 2 `<span/>`, changing the DOM tree.
It's not possible in stable, clients are already starting in Odoo 16.4 atm, so 16.3 is really not possible IMO.
Unless @qsm-odoo has another opinion?

**Review by rdeodoo — COMMENTED** [review]
Working as expected but DOM modification in 16.3, see review point

File: addons/website/static/src/systray_items/translate_website.xml (hunk 10 above)
Comment (qsm-odoo) [reply addons/website/static/src/systray_items/translate_website.xml:7]: @rdeodoo it's in static xml, so it should be ok to add a new node :wink:

File: addons/website/static/src/components/navbar/navbar.js (hunk 6 above)
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.js:47]: Changed

File: addons/website/static/src/components/navbar/navbar.xml (hunk 5 above)
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.xml]: Squashed & used `add=` to avoid anything possibly broken

File: addons/website/static/src/components/navbar/navbar.scss (hunk 7 above)
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss]: Line removed at the end

File: addons/website/static/src/components/navbar/navbar.scss (hunk 8 above)
Comment (rdeodoo) [reply addons/website/static/src/components/navbar/navbar.scss]: Fixed by sri

**Comment by rdeodoo** [comment]
@qsm-odoo @stefanorigano Ty, I squashed everything and rewrote the commit msg + adapted the last points.
Everything should be good now?

File: addons/website/static/src/systray_items/edit_website.scss [hunk 11]
```diff
@@ -1,3 +1,8 @@
 .o_edit_website_container {
     margin: -$o-navbar-padding-v 0;
+
+    &.o_not_translatable .o_nav_entry, &.o_not_translatable .o_nav_entry:hover {
+        background-color: inherit;
+        color: inherit;
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/systray_items/edit_website.scss:7]:
```suggestion
    &.o_not_translatable .o_nav_entry {
        &, &:hover {
            background-color: inherit;
            color: inherit;
        }
    }
```

File: addons/website/static/src/systray_items/edit_website.xml [hunk 12]
```diff
@@ -1,11 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <templates xml:space="preserve">
 <t t-name="website.EditWebsiteSystray" owl="1">
-    <div class="o_edit_website_container d-none d-md-block" t-att-class="{ o_translatable: translatable }">
-        <a href="#" class="h-100" accesskey="a" t-on-click="startEdit" t-att-class="{'btn btn-primary d-flex align-items-center rounded-0 border-0 fw-normal': !translatable, 'o_nav_entry': translatable}">
-            <span t-if="translatable" class="fa fa-pencil me-2"/>
-            <span t-else="" class="fa me-2" t-att-class="{ 'fa-refresh fa-spin': this.state.isLoading, 'fa-pencil': !this.state.isLoading }"/>
-            <t t-esc="label"/>
+    <div class="o_menu_systray_item o_edit_website_container d-none d-md-block" t-att-class="{ o_translatable: translatable, o_not_translatable: !translatable }">
```
Comment (qsm-odoo) [inline addons/website/static/src/systray_items/edit_website.xml:4]: Do we need the new `o_not_translatable` class? Seems like it is used for the CSS rule only, so why not `:not(.o_translatable)`? :thinking: I am not a fan of `:not(` in general but I prefer it over duplicating classes to do the `not` :wink:

File: addons/website/static/src/systray_items/translate_website.scss [hunk 13]
```diff
@@ -1,3 +1,8 @@
 .o_translate_website_container {
     margin: -$o-navbar-padding-v 0;
+
+    .btn .o_nav_entry, .btn .o_nav_entry:hover {
+        background-color: inherit;
+        color: inherit;
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/systray_items/translate_website.scss:7]: Same here

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Sorry, took me long enough to have a look. Details but...

File: addons/website/static/src/systray_items/translate_website.scss (hunk 13 above)
Comment (rdeodoo) [reply addons/website/static/src/systray_items/translate_website.scss]: Changed

File: addons/website/static/src/systray_items/edit_website.scss (hunk 11 above)
Comment (rdeodoo) [reply addons/website/static/src/systray_items/edit_website.scss]: Changed

File: addons/website/static/src/systray_items/edit_website.xml (hunk 12 above)
Comment (rdeodoo) [reply addons/website/static/src/systray_items/edit_website.xml]: `o_translatable` is actually dead code.
I noticed that already when reviewing but :shrug: 
Only kept `o_not_translatable`, no need of `:not` as a bonus

**Review by rdeodoo — COMMENTED** [review]
Adapted

**Comment by qsm-odoo** [comment]
> you may want to rebuild or fix this PR as it has failed CI.

Don't know what's going on, but 500 commits late so I rebased, hopefully this will solve the issue.

> Still broken in enterprise with the same thing I reported before: you can click the publish button by clicking on its right where the > cursor indicates it is not clicking on the button + there is an ugly gap where clicking does nothing.
> 
> But I guess we can merge this first:

@robodoo r+

**Comment by qsm-odoo** [comment]
> you may want to rebuild or fix this PR as it has failed CI.

Tests were red for a good reason, [fixed it](https://github.com/odoo/odoo/compare/e6511ab7122627887553ce8dd83d8459eb883527..afad48dee53ff3dd55b369a69adb6e1fbd38ac33).

> Still broken in enterprise with the same thing I reported before: you can click the publish button by clicking on its right where the > cursor indicates it is not clicking on the button + there is an ugly gap where clicking does nothing.
> 
> But I guess we can merge this first:

@robodoo r+
